ext/flexmailer/src/ClickTracker/HtmlClickTracker.php

```diff
@@ -51,9 +51,9 @@ public static function replaceHrefUrls($html, $replace) {
 
     // Find anything like href="..." or href='...' inside a tag.
     $tmp = preg_replace_callback(
-      ';(\<a[^>]*href *= *")([^">]+)(");i', $callback, $html);
+      ';(\<a[^>]*href *= *")([^">]+)(");iu', $callback, $html);
     return preg_replace_callback(
-      ';(\<a[^>]*href *= *\')([^\'>]+)(\');i', $callback, $tmp);
+      ';(\<a[^>]*href *= *\')([^\'>]+)(\');iu', $callback, $tmp);
```

Comment:
so the u is for 'unicode'?

Reply:
yes AFAIK

Reply:
I had run into some problems using `u` when the string is not in fact encoded in UTF8 but contains non-ascii chars - the preg_replace can result in the empty string when the resulting string can't be represented as a valid utf8 string: https://lab.civicrm.org/dev/core/-/issues/2127#note_49410. That might not be a problem here though, since it looks like this is appending not replacing.

Reply:
Learn something new everyday... :)

> That might not be a problem here though, since it looks like this is appending not replacing.

If I'm reading correctly, another difference here is that dev/core#2127 is importing CSV files (ie *the string comes from external/untrusted files*), so you can have different inputs where similar content is encoded differently (eg latin1 `a0` vs unicode `c2a0` in your comment above). With #21597, content comes from `civicrm_mailing.body_html`, so it should be Unicode (utf8/utf8mb4). If `civicrm_mailing.body_html` is not Unicode, then there's a bug somewhere else (eg data-input or data-storage).